### PR TITLE
uaa suport  ROPC(resource owner password credentials) grant of OAuth2

### PR DESCRIPTION
**Pull request by anjia0532**
fix jhipster/generator-jhipster#6227 and remove _OAuth2InterceptedFeignConfiguration.java useless packages.

- Please make sure the below checklist is followed for Pull Requests.

- [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [ ] Tests are added where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by anjia0532** [comment]
ping @deepu105

**Comment by anjia0532** [comment]
may be  bast way in the netflix doc ([Cookies and Sensitive Headers](http://cloud.spring.io/spring-cloud-netflix/spring-cloud-netflix.html#_cookies_and_sensitive_headers))  setting `zuul.sensitiveHeaders` as empty string (jhipster-registry's central-config application.yml).but it not work.

**Comment by anjia0532** [comment]
@PierreBesson yup
gateway's application.yml
```yml
zuul: 
    sensitiveHeaders:  
```
the cookie and auth header not pass

**Comment by xetys** [comment]
The issue you name is, that when a microservice calls another by doing a client credentials grant, obviously no information about the requesting user is passed. This is intended, as it opens you to specify distinct policies for user-to-service and service-to-service communication. For the latter case, the `@AuthorizedFeignClient` helps to do this according to the CC grant of OAuth2. 

What you want to do is token forwarding. For this purpose, there is no need to implement `OAuth2FeignRequestInterceptor` as your feign client just do header modification, while the implented interceptor performs a CC flow.

However, you point out an issue I noticed earlier in development: there is need for both scenarios. I would suggest to

* let your configuration class implement just `RequestInterceptor`, but name the class more precisely `OAuth2UserClientFeignConfiguration`
* keep the `OAuth2InterceptedFeignConfiguration` but name it more precisly like `OAuth2ServiceClientFeignConfiguration`
* rename `AuthorizedFeignClient` to `AuthorizedServiceFeignCleint` and apply the `OAuth2ServiceClientFeignConfiguration` on this
* create a `AuthorizedUserFeignClient` and apply ``OAuth2UserClientFeignConfiguration` here


the result is, that a user now can decide, if the second call is done in behalf of the user or the microservice itself.

Maybe it makes sense to move the annotations and configuration to the jhipster library to reduce generated code

**Comment by anjia0532** [comment]
Is there anything I can do?

**Comment by xetys** [comment]
You can do the points I suggested, if you feel confident. If you run into issues, don't hesitate to ask me.

**Comment by anjia0532** [comment]
how to get `Authorization` header in the `RequestInterceptor` ?

in my dev project, i'll got this header(oracle jdk 8), cannot get this when i publish this server to docker (openjdk 8 alpine) .

```
tree .
. 
├── AuthorizedServiceFeignCleint.java
├── AuthorizedUserFeignClient.java
├── OAuth2ServiceClientFeignConfiguration.java
├── OAuth2UserClientFeignConfiguration.java
└── UserFeignClientInterceptor.java
```

**Comment by cbornet** [comment]
@anjia0532 can you change the PR title to something more explicit ?

**Comment by anjia0532** [comment]
@cbornet yup

**Comment by xetys** [comment]
@anjia0532 honestly, I don't get your question...you already built the auth header logic in your `FeignClientInterceptor` class. There is just no need to implement the OAuth2RequestIntercetor. 

The docker problem, I just don't understand the issue, can you explain it more precisely?

**Comment by anjia0532** [comment]
@xetys  `_AuthorizedFeignClient.java` `_OAuth2InterceptedFeignConfiguration.java` is CC flow.
`_AuthorizedUserFeignClient.java` `_OAuth2UserClientFeignConfiguration.java` `_UserFeignClientInterceptor.java` is ROPC flow.

i did not rename `_AuthorizedFeignClient` and `_OAuth2InterceptedFeignConfiguration`,  it needs a lots of changes to doc and code.

**Comment by anjia0532** [comment]
ping @xetys , ref https://github.com/spring-cloud/spring-cloud-netflix/issues/1330
it must be share securityContent when the hystrix exec. if not ,`UserFeignClientInterceptor` cannot got securityContent

```yml
feign:
  hystrix:
    enabled: true

hystrix:
  shareSecurityContext: true
```

File: generators/server/templates/src/main/java/package/client/_OAuth2InterceptedFeignConfiguration.java [hunk 1]
```diff
@@ -41,6 +38,6 @@ public OAuth2InterceptedFeignConfiguration(LoadBalancedResourceDetails loadBalan
 
     @Bean(name = "oauth2RequestInterceptor")
     public RequestInterceptor getOAuth2RequestInterceptor() throws IOException {
-        return new OAuth2FeignRequestInterceptor(new DefaultOAuth2ClientContext(), loadBalancedResourceDetails);
+        return new FeignClientInterceptor(new DefaultOAuth2ClientContext(), loadBalancedResourceDetails);
```
Comment (xetys) [inline generators/server/templates/src/main/java/package/client/_OAuth2InterceptedFeignConfiguration.java]: This change looks wrong, shouldn't that use the OAuth2FeignRequestInterceptor from spring like before?

Comment (anjia0532) [reply generators/server/templates/src/main/java/package/client/_OAuth2InterceptedFeignConfiguration.java]: sorry, it was old code.

File: generators/server/templates/src/main/resources/config/_application.yml [hunk 2]
```diff
@@ -88,6 +88,13 @@ hystrix:
 #                isolation:
 #                    thread:
 #                        timeoutInMilliseconds: 10000
+# See https://github.com/spring-cloud/spring-cloud-netflix/issues/1330
+feign:
+  hystrix:
```
Comment (jdubois) [inline generators/server/templates/src/main/resources/config/_application.yml]: Can you use the same indentation as the rest of the file? 4 spaces, please

**Comment by jdubois** [comment]
This looks like a good PR, but we have failing CI, a pending review and a conflict -> I don't think they are complex to solve, @anjia0532 do you have time to do it? Do you need help?

**Comment by anjia0532** [comment]
@jdubois the ci error log `[ERROR] Failed to execute goal org.apache.maven.plugins:maven-war-plugin:2.6:war (default-war) on project jhipster-sample-application: Execution default-war of goal org.apache.maven.plugins:maven-war-plugin:2.6:war failed: Plugin org.apache.maven.plugins:maven-war-plugin:2.6 or one of its dependencies could not be resolved: Failed to collect dependencies at org.apache.maven.plugins:maven-war-plugin:jar:2.6 -> org.codehaus.plexus:plexus-io:jar:2.4.1: Failed to read artifact descriptor for org.codehaus.plexus:plexus-io:jar:2.4.1: Could not transfer artifact org.codehaus.plexus:plexus-io:pom:2.4.1 from/to central (https://repo.maven.apache.org/maven2): Connection reset -> [Help 1]` , i don't change pom.xml file.

**Comment by jdubois** [comment]
@anjia0532 might be a CI error, let me restart the build

**Comment by anjia0532** [comment]
@jdubois Cheers! Long live rebuild!

**Comment by deepu105** [comment]
@anjia0532 can you fix the conflict. we can try to merge this
@jdubois are you ok with this? its been open for a while now

**Comment by anjia0532** [comment]
@deepu105  Travis tests are green.

**Comment by deepu105** [comment]
yes but seems there is a merge conflict in `generators/server/templates/src/main/java/package/client/_OAuth2InterceptedFeignConfiguration.java` can you check

**Comment by anjia0532** [comment]
@deepu105 sorry, i don't have write access and it's has too many diff between 2 branch.

https://github.com/jhipster/generator-jhipster/blob/master/generators/server/templates/src/main/java/package/client/_OAuth2InterceptedFeignConfiguration.java

https://github.com/anjia0532/generator-jhipster/blob/master/generators/server/templates/src/main/java/package/client/_OAuth2InterceptedFeignConfiguration.java

**Comment by deepu105** [comment]
how come you dont have write access to your own fork/branch?

@jdubois since you are a better person for this I'll leave this to you

**Comment by cbornet** [comment]
FIrst thing is that changes should have been done in a branch

**Comment by jdubois** [comment]
OK I'm doing the merge manually

**Comment by anjia0532** [comment]
@jdubois Sorry for the delay and thanks : )

**Pull request merged**